### ISSUE TITLE
Go apps affecting projects incorrectly

### ISSUE THREAD

**Issue by raoulkent** [nx-go/nx-go#134]
### Current Behavior

Currently, whenever there is a change by creating or deleting an nx-go application or library other `nx/node` apps are affected. This was observed in a repo using both `nx/node` as well as `nx-go` apps, where deleting a unused `nx-go` app caused the pipeline to output the `nx/node` app as affected.

I tried replicating this by:

1. Starting a new repo from scratch, with `npx create-nx-workspace go-workspace --preset=@nx-go/nx-go`
2. Adding nx/node and other required plugins

### Expected Behavior

Using the same example as in the current behaviour:

### First example

1. Add nx/node app (named `foo`)
2. Commit -> `nx show projects affected` outputs none
3. Add nx-go app (named `spam`)
4. `nx show projects affected` *SHOULD* output `spam`

### First example

1. Add nx/node app (named `foo`)
2. Commit -> `nx show projects affected` outputs none
3. Add nx-go app (named `spam`)
4. `nx show projects affected` outputs both `foo` and `spam`

### Second example

1. Add nx/node app (named `foo`)
2. Commit -> `nx show projects affected` outputs none
3. Add nx-go app (named `spam`) and commit
4. Remove `spam`
5. `nx show projects affected` outputs `foo`

### Nx Report

```shell
NX   Report complete - copy this into the issue template

Node   : 21.7.3
OS     : darwin-arm64
npm    : 10.8.2

nx (global)        : 19.0.7
nx                 : 19.0.7
@nx/js             : 19.0.7
@nx/jest           : 19.0.7
@nx/linter         : 19.0.7
@nx/eslint         : 19.0.7
@nx/workspace      : 19.0.7
@nx/devkit         : 19.0.7
@nx/esbuild        : 19.0.7
@nx/eslint-plugin  : 19.0.7
@nx/node           : 19.0.7
@nrwl/tao          : 19.0.7
typescript         : 5.4.5
---------------------------------------
Registered Plugins:
@nx/eslint/plugin
@nx/jest/plugin
@nx-go/nx-go
---------------------------------------
Community plugins:
@nx-go/nx-go : 3.0.0
```

### nx.json

```json
{
  "$schema": "./node_modules/nx/schemas/nx-schema.json",
  "namedInputs": {
    "default": [
      "{projectRoot}/**/*",
      "sharedGlobals"
    ],
    "production": [
      "default",
      "!{projectRoot}/.eslintrc.json",
      "!{projectRoot}/eslint.config.js",
      "!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)",
      "!{projectRoot}/tsconfig.spec.json",
      "!{projectRoot}/jest.config.[jt]s",
      "!{projectRoot}/src/test-setup.[jt]s",
      "!{projectRoot}/test-setup.[jt]s"
    ],
    "sharedGlobals": [
      "{workspaceRoot}/go.work"
    ]
  },
  "targetDefaults": {
    "@nx/esbuild:esbuild": {
      "cache": true,
      "dependsOn": [
        "^build"
      ],
      "inputs": [
        "production",
        "^production"
      ]
    }
  },
  "plugins": [
    {
      "plugin": "@nx/eslint/plugin",
      "options": {
        "targetName": "lint"
      }
    },
    {
      "plugin": "@nx/jest/plugin",
      "options": {
        "targetName": "test"
      }
    },
    "@nx-go/nx-go"
  ],
  "neverConnectToCloud": true
}
```

### Go Version

go version go1.23.0 darwin/arm64

**Comment by raoulkent** [comment]
I'm happy to help, but I'm rather new to the internals of Nx and more so Nx-go.
If you do not have the time to analyse/create a fix, please tell me where I might start that process.

**Comment by johnmckenna** [comment]
Hi @raoulkent you'll see `sharedGlobals` in your `nx.json` as this is a go workspace then I presume the design is to assume all projects are for go and therefore `go.work` changes affects all projects.

So even though `foo` is not a go app, it will be affected due to the `sharedGlobals`.

Not 100% sure it's a bug @utarwyn can comment on that.

You should see that if you remove `sharedGlobals` from the default `namedInputs` and commit that (because `nx.json` also affects all projects) that then only `spam` is affected even though `go.work` is still modified.

Alternatively, leave `nx.json` as is and remove the modification to `go.work` and you will also see only `spam` is affected.

You could create an nx workspace and then add nx go and nx node, you'll still have to work out your `namedInputs` though for each type of project reference on this is here: https://nx.dev/reference/inputs

**Comment by raoulkent** [comment]
Hi @johnmckenna!
Thank you for the thoughtful response, that's very insightful. It's much appreciated that you laid them out with such clarity. I'll dig into the details and see how best to make it work for our use case.

If this is not a bug, it does raise the question of how to convey that to a new user. I wasn't aware of the `sharedGlobals` and how that might affect things.
In our case, we're trying to leverage `affected` as much as possible. So when I discovered this issue it put a spanner in the works of developing a lean CICD pipeline for our polyglot repo!

**Comment by utarwyn** [comment]
Hello @raoulkent !
I'm sorry, I came really late to the discussion. I took some time to test this behavior on my side, and it's true that it's not optimal at all. Especially when you have a lot of different types of projects. The aim of the plugin is to work with **all types of workspaces**, whether they only contain Go projects or any other type.

The most important thing for nx-go is to know what the dependencies are and which projects communicate with each other. After a lot of testing and thinking... I don't understand why I put `go.work` in the list of sharedGlobals. It doesn't really make sense, because when a project is added or deleted, it's automatically “ affected ”. And the graph system already knows which projects are using it, so it can “ affect ” them too. It's as simple as that.

So my idea would be to simply remove this sharedGlobals, and provide a migration so that all workspaces are rid of it.
What do you think of it @phillipCouto? @johnmckenna?

**Comment by phillipCouto** [comment]
> Hello @raoulkent ! I'm sorry, I came really late to the discussion. I took some time to test this behavior on my side, and it's true that it's not optimal at all. Especially when you have a lot of different types of projects. The aim of the plugin is to work with **all types of workspaces**, whether they only contain Go projects or any other type.
> 
> The most important thing for nx-go is to know what the dependencies are and which projects communicate with each other. After a lot of testing and thinking... I don't understand why I put `go.work` in the list of sharedGlobals. It doesn't really make sense, because when a project is added or deleted, it's automatically “ affected ”. And the graph system already knows which projects are using it, so it can “ affect ” them too. It's as simple as that.
> 
> So my idea would be to simply remove this sharedGlobals, and provide a migration so that all workspaces are rid of it. What do you think of it @phillipCouto? @johnmckenna?

I agree that it is a little heavy handed for changes to `go.work` to impact the entire workspace but changes to `go.work` can impact every Go project in the workspace. 

Due to the `toolchain` and `replace` directives maybe we should add `go.work` to the default inputs for targets generated by `nx-go`? 

Or maybe we add a named input that can be easily used by go project targets that in the future can be easily adjusted by `nx-go` if need be. We may want to include `go.work.sum` as an example or include root level `go.mod` and `go.sum` for workspaces that do not use go workspaces.

Thoughts?

**Comment by utarwyn** [comment]
@phillipCouto

Thank you for this valuable feedback. I'd actually forgotten that these files can contain the Go version, as well as other information such as `toolchain` and `replace` directives. Projects managed by **nx-go** must therefore be dependent on their own `go.mod` file, or on the workspace's `go.work`/`go.mod` file. But other projects should not be dependent on it - that's the complexity of the subject.

I'm going to think about a solution, and I'm open to any idea!

**Comment by techfg** [comment]
@utarwyn -

First off, thanks for your work on this plugin, greatly appreciated!

Regarding `go.work`, agreed that it should not be included in `sharedGlobals` as it will affect any project that references `sharedGlobals` in its [inputs](https://nx.dev/reference/inputs).  However, as @phillipCouto and you both mention, it is important to have `go.work` included in `go` projects as an input.  Ironically, even though `go.work` is in `sharedGlobals` by default, because `@nx-go` projects (apps & libs) do not define inputs in the default generated output, if you modify `go.work` and run `nx shows projects --affected`, `go` projects will not be listed.  Additionally, `go.work.sum` is not considered which is also an asset that could change and needs to trigger tasks when modified.

For example:

1. `git clone https://github.com/techfg/nx-go-deps-repro.git`
2. `cd nx-go-deps-repro`
3. `npm install`
4. `npx nx run-many -t build`
5. `npx nx show projects --affected`
6. Add a comment to ./go.work to simulate a modification (e.g., `// my test comment`)
7. `npx nx show projects --affected`

You'll see in step 5, no projects were listed (as expected), however in step 7, `reactapp` was listed but `goapp` was not (not expected).  

In order to resolve and ensure proper inputs/outputs, the following seems to work for me:

1. Define `golang` namedInput which can be used by all tasks.  Note that namedInput isn't required, the items could be included explicitly in each task but using named inputs simplifies config.  This includes `go.work` and `go.work.sum` along with project level `go` related files.
2. Enable caching (not required but beneficial to overall affected detection)
3. Define inputs for each task that includes "golang", "sharedGlobals" and all "go" files
4. Define output for build task (helps with affected detection)
 
```json nx.json
...
  "namedInputs": {
    "default": [
      "{projectRoot}/**/*",
      "sharedGlobals"
    ],
    "production": [
      "default",
      "!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)",
      "!{projectRoot}/tsconfig.spec.json",
      "!{projectRoot}/src/test-setup.[jt]s"
    ],
    "sharedGlobals": [
      "{workspaceRoot}/.github/workflows/ci.yml"
    ],
    "golang": [
      "{projectRoot}/go.mod",
      "{projectRoot}/go.sum",
      "{projectRoot}/**/*.go",      
      "{workspaceRoot}/go.work",
      "{workspaceRoot}/go.work.sum"
    ]
  },
  "targetDefaults": {
    "@nx-go/nx-go:build": {
      "cache": true,
      "inputs": [
        "golang",
        "^production",
        "sharedGlobals"
      ],
      "outputs": [
        "{options.outputPath}"
      ]
    },
    "@nx-go/nx-go:test": {
      "cache": true,
      "inputs": [
        "golang",
        "^production",
        "sharedGlobals"
      ]
    },
    "@nx-go/nx-go:lint": {
      "cache": true,
      "inputs": [
        "golang",
        "^production",
        "sharedGlobals"
      ]
    }
  },
...
```

To see the result of the above changes:

1. `git checkout fix/configure-nxgo-tasks`
2. `npx nx reset`
3. `npx nx show projects --affected --base=fix/configure-nxgo-tasks`
4. Add a comment to ./go.work to simulate a modification (e.g., `// my test comment`)
5. `npx nx show projects --affected --base=fix/configure-nxgo-tasks`

You'll see in step 3, no projects were listed (as expected) and now in Step 5, `goapp` was listed but `reactapp` was not (as expected).

In short, I think there are several improvement/adjustments that can be made to the "default" @nx-go configurations that are generated.  Since this issue is targeting specifically how `sharedGlobals` is incorrectly affecting non-go projects, I'll create a separate issue to track the broader list.  In the meantime, specific to `sharedGlobals`, `go.work` I believe `go.work` should be removed and instead, be a task specific include for each go project task that could be affected by a change to it.

Thank you!
